### PR TITLE
Explicit checks for ZST + null fallbacks

### PR DESCRIPTION
**Pull request by elichai**
This is based off a suggestion by @real-or-random https://github.com/rust-bitcoin/rust-secp256k1/pull/145#issuecomment-523021675

This PR is a combination of #144 and #145 but with one difference, calling `as_c_ptr()` on an empty slice will result in a null pointer.

The point is to have explicit checks for empty slices, but if somewhere we miss a check then this will make sure a null is produced which will be caught by `ARG_CHECK` in upstream and make the code panic.

### CODE REVIEW

File: src/ffi.rs [hunk 1]
```diff
@@ -366,6 +366,37 @@ unsafe fn strlen(mut str_ptr: *const c_char) -> usize {
 }
 
 
+/// A trait for producing pointers that will always be valid in C.
+/// Rust doesn't promise what pointers does it give to ZST (https://doc.rust-lang.org/nomicon/exotic-sizes.html#zero-sized-types-zsts)
+/// This trait will give a pointer to a static variable in case the type is empty.
```
Comment (real-or-random) [inline src/ffi.rs]: nit: comment should be updated to mention the NULL thing (and maybe why we prefer NULL)

Comment (real-or-random) [reply src/ffi.rs]: ACK b7e20c5b120c42769872eb116ab4dbed2d965a2c

**Comment by apoelstra** [comment]
Can you add a unit test that passing a zero-width slice to e.g. `SecretKey::from_slice` returns an error.

I think this existing code will trigger an abort in libsecp.

**Comment by elichai** [comment]
That actually doesn't trigger abort because `SecretKey::from_slice` is written really weirdly.
it matches over the length of data and calls libsecp only if it matches `constants::SECRET_KEY_SIZE`.

I'll add that test anyway, and maybe also refactor that function to have a simple if instead of that match heh :)

**Comment by apoelstra** [comment]
Yeah, sounds like a copy of the `PublicKey::from_slice` logic.

**Comment by apoelstra** [comment]
I think you might as well leave the original logic in place, if it works.

Also can you change the commit message :)
